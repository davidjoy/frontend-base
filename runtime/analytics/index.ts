export {
  configureAnalytics,
  getAnalyticsService,
  identifyAnonymousUser,
  identifyAuthenticatedUser,
  resetAnalyticsService,
  sendPageEvent,
  sendTrackEvent,
  sendTrackingLogEvent
} from './interface';
export { default as MockAnalyticsService } from './MockAnalyticsService';
export { default as SegmentAnalyticsService } from './SegmentAnalyticsService';
