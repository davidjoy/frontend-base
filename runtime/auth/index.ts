export { default as AxiosJwtAuthService } from './AxiosJwtAuthService';
export {
  AUTHENTICATED_USER_CHANGED,
  AUTHENTICATED_USER_TOPIC,
  configureAuth,
  ensureAuthenticatedUser,
  fetchAuthenticatedUser,
  getAuthenticatedHttpClient,
  getAuthenticatedUser,
  getAuthService,
  getHttpClient,
  getLoginRedirectUrl,
  getLogoutRedirectUrl,
  hydrateAuthenticatedUser,
  redirectToLogin,
  redirectToLogout,
  setAuthenticatedUser
} from './interface';
export { default as MockAuthService } from './MockAuthService';
