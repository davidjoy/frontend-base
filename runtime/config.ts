/**
 * #### Import members from **@edx/frontend-base**
 *
 * The configuration module provides utilities for working with an application's configuration
 * document (ConfigDocument).  Configuration variables can be supplied to the
 * application in three different ways.  They are applied in the following order:
 *
 * - Site Configuration File (site.config.tsx)
 * - Initialization Config Handler
 * - Runtime Configuration
 *
 * Last one in wins, and are deep merged together.  Variables with the same name defined via the
 * later methods will override any defined using an earlier method.  i.e., if a variable is defined
 * in Runtime Configuration, that will override the same variable defined in either of the earlier
 * methods.  Configuration defined in a JS file will override any default values below.
 *
 * ##### Site Configuration File
 *
 * Configuration variables can be supplied in a file named site.config.tsx.  This file must
 * export either an Object containing configuration variables or a function.  The function must
 * return an Object containing configuration variables or, alternately, a promise which resolves to
 * an Object.
 *
 * Using a function or async function allows the configuration to be resolved at runtime (because
 * the function will be executed at runtime).  This is not common, and the capability is included
 * for the sake of flexibility.
 *
 * The Site Configuration File is well-suited to extensibility use cases or component overrides,
 * in that the configuration file can depend on any installed JavaScript module.  It is also the
 * preferred way of doing build-time configuration if runtime configuration isn't used by your
 * deployment of the platform.
 *
 * Exporting a config object:
 * ```
 * const config = {
 *   LMS_BASE_URL: 'http://localhost:18000'
 * };
 *
 * export default config;
 * ```
 *
 * Exporting a function that returns an object:
 * ```
 * function getConfig() {
 *   return {
 *     LMS_BASE_URL: 'http://localhost:18000'
 *   };
 * }
 * ```
 *
 * Exporting a function that returns a promise that resolves to an object:
 * ```
 * function getAsyncConfig() {
 *   return new Promise((resolve, reject) => {
 *     resolve({
 *       LMS_BASE_URL: 'http://localhost:18000'
 *     });
 *   });
 * }
 *
 * export default getAsyncConfig;
 * ```
 *
 * ##### Initialization Config Handler
 *
 * The configuration document can be extended by
 * applications at run-time using a `config` initialization handler.  Please see the Initialization
 * documentation for more information on handlers and initialization phases.
 *
 * ```
 * initialize({
 *   handlers: {
 *     config: () => {
 *       mergeConfig({
 *         CUSTOM_VARIABLE: 'custom value',
 *         LMS_BASE_URL: 'http://localhost:18001' // You can override variables, but this is uncommon.
  *       }, 'App config override handler');
 *     },
 *   },
 * });
 * ```
 *
 * ##### Runtime Configuration
 *
 * Configuration variables can also be supplied using the "runtime configuration" method, taking
 * advantage of the Micro-frontend Config API in edx-platform. More information on this API can be
 * found in the ADR which introduced it:
 *
 * https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/mfe_config_api/docs/decisions/0001-mfe-config-api.rst
 *
 * The runtime configuration method can be enabled by supplying a MFE_CONFIG_API_URL via one of the other
 * two configuration methods above.
 *
 * Runtime configuration is particularly useful if you need to supply different configurations to
 * a single deployment of a micro-frontend, for instance.  It is also a perfectly valid alternative
 * to build-time configuration, though it introduces an additional API call to edx-platform on MFE
 * initialization.
 *
 *
 * @module Config
 */

import merge from 'lodash.merge';
import {
  AppConfigTypes, ApplicationModuleConfig, ConfigurableAppConfig, EnvironmentTypes, SiteConfig
} from '../types';
import { CONFIG_CHANGED } from './constants';
import { publish } from './subscriptions';

let config: SiteConfig = {
  ACCESS_TOKEN_COOKIE_NAME: 'edx-jwt-cookie-header-payload',
  CSRF_TOKEN_API_PATH: '/csrf/api/v1/token',
  ENVIRONMENT: EnvironmentTypes.PRODUCTION,
  IGNORED_ERROR_REGEX: null,
  LANGUAGE_PREFERENCE_COOKIE_NAME: 'openedx-language-preference',
  PUBLIC_PATH: '/',
  REFRESH_ACCESS_TOKEN_API_PATH: '/login_refresh',
  USER_INFO_COOKIE_NAME: 'edx-user-info',
  MFE_CONFIG_API_URL: null,

  SEGMENT_KEY: null,
  SUPPORT_EMAIL: null,

  // Optional Frontends
  ORDER_HISTORY_URL: null,
  SUPPORT_URL: null,
  TERMS_OF_SERVICE_URL: null,
  PRIVACY_POLICY_URL: null,
  ACCESSIBILITY_URL: null,

  // Optional Backends
  CREDENTIALS_BASE_URL: null,
  DISCOVERY_API_BASE_URL: null,
  ECOMMERCE_BASE_URL: null,
  PUBLISHER_BASE_URL: null,

  apps: {},
  pluginSlots: {},
  custom: {},

  APP_ID: '',
  BASE_URL: '',
  SITE_NAME: '',

  // Frontends
  ACCOUNT_PROFILE_URL: '',
  ACCOUNT_SETTINGS_URL: '',
  LEARNER_DASHBOARD_URL: '',
  LEARNING_BASE_URL: '',
  LOGIN_URL: '',
  LOGOUT_URL: '',
  MARKETING_SITE_BASE_URL: '',

  // Backends
  LMS_BASE_URL: '',
  STUDIO_BASE_URL: '',

  // Branding
  FAVICON_URL: '',
  LOGO_TRADEMARK_URL: '',
  LOGO_URL: '',
  LOGO_WHITE_URL: ''
};

/**
 * Getter for the application configuration document.  This is synchronous and merely returns a
 * reference to an existing object, and is thus safe to call as often as desired.
 *
 * Example:
 *
 * ```
 * import { getConfig } from '@openedx/frontend-base';
 *
 * const {
 *   LMS_BASE_URL,
 * } = getConfig();
 * ```
 *
 * @returns {ConfigDocument}
  */
export function getConfig() {
  return config;
}

/**
 * Replaces the existing ConfigDocument.  This is not commonly used, but can be helpful for tests.
 *
 * Example:
 *
 * ```
 * import { setConfig } from '@openedx/frontend-base';
 *
 * setConfig({
 *   LMS_BASE_URL, // This is overriding the ENTIRE document - this is not merged in!
 * });
 * ```
 *
 * @param {ConfigDocument} newConfig
 */
export function setConfig(newConfig: SiteConfig) {
  config = newConfig;
  publish(CONFIG_CHANGED);
}

/**
 * Merges additional configuration values into the site config returned by `getConfig`.  Will
 * override any values that exist with the same keys.
 *
 * ```
 * mergeConfig({
 *   NEW_KEY: 'new value',
 *   OTHER_NEW_KEY: 'other new value',
 * });
 *
 * This function uses lodash.merge internally to merge configuration objects
 * which means they will be merged recursively.  See https://lodash.com/docs/latest#merge for
 * documentation on the exact behavior.
 *
 * @param {Object} newConfig
 */
export function mergeConfig(newConfig: Partial<SiteConfig>) {
  config = merge(config, newConfig);
  publish(CONFIG_CHANGED);
}

export function patchAppModuleConfig(appId: string, appModuleConfig: ApplicationModuleConfig) {
  if (config.apps[appId] !== undefined) {
    const app = config.apps[appId];
    if (app.type === AppConfigTypes.INTERNAL || app.type === AppConfigTypes.FEDERATED) {
      const configurableApp = app as ConfigurableAppConfig;
      configurableApp.config = appModuleConfig;
    }
  }
}

/**
 * An object describing the current application configuration.
 *
 * In its most basic form, this document contains some sensible defaults. The initialization
 * process then merges additional configuration into this document using the configuration methods
 * described above. (Site configuration file, initialization config handler, and runtime
 * configuration)
 *
 * @name ConfigDocument
 * @memberof module:Config
 * @property {string} ACCESS_TOKEN_COOKIE_NAME
 * @property {string} ACCOUNT_PROFILE_URL
 * @property {string} ACCOUNT_SETTINGS_URL
 * @property {string} BASE_URL The URL of the current application.
 * @property {string} CREDENTIALS_BASE_URL
 * @property {string} CSRF_TOKEN_API_PATH
 * @property {string} DISCOVERY_API_BASE_URL
 * @property {string} PUBLISHER_BASE_URL
 * @property {string} ECOMMERCE_BASE_URL
 * @property {string} ENVIRONMENT This is one of: development, production, or test.
 * @property {string} IGNORED_ERROR_REGEX
 * @property {string} LANGUAGE_PREFERENCE_COOKIE_NAME
 * @property {string} LEARNING_BASE_URL
 * @property {string} LMS_BASE_URL
 * @property {string} LOGIN_URL
 * @property {string} LOGOUT_URL
 * @property {string} STUDIO_BASE_URL
 * @property {string} MARKETING_SITE_BASE_URL
 * @property {string} ORDER_HISTORY_URL
 * @property {string} REFRESH_ACCESS_TOKEN_API_PATH
 * @property {string} SEGMENT_KEY
 * @property {string} SITE_NAME
 * @property {string} USER_INFO_COOKIE_NAME
 * @property {string} LOGO_URL
 * @property {string} LOGO_TRADEMARK_URL
 * @property {string} LOGO_WHITE_URL
 * @property {string} FAVICON_URL
 * @property {string} MFE_CONFIG_API_URL
 * @property {string} APP_ID
 * @property {string} SUPPORT_URL
 */
