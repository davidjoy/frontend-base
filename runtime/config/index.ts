/**
 * #### Import members from **@edx/frontend-base**
 *
 * The configuration module provides utilities for working with an application's configuration
 * document (SiteConfig).  Configuration variables can be supplied to the
 * application in three different ways.  They are applied in the following order:
 *
 * - Site Configuration File (site.config.tsx)
 * - Initialization Config Handler
 * - Runtime Configuration
 *
 * Last one in wins, and are deep merged together.  Variables with the same name defined via the
 * later methods will override any defined using an earlier method.  i.e., if a variable is defined
 * in Runtime Configuration, that will override the same variable defined in either of the earlier
 * methods.  Configuration defined in a JS file will override any default values below.
 *
 * ##### Site Configuration File
 *
 * Configuration variables can be supplied in a file named site.config.tsx.  This file must
 * export either an Object containing configuration variables or a function.  The function must
 * return an Object containing configuration variables or, alternately, a promise which resolves to
 * an Object.
 *
 * Using a function or async function allows the configuration to be resolved at runtime (because
 * the function will be executed at runtime).  This is not common, and the capability is included
 * for the sake of flexibility.
 *
 * The Site Configuration File is well-suited to extensibility use cases or component overrides,
 * in that the configuration file can depend on any installed JavaScript module.  It is also the
 * preferred way of doing build-time configuration if runtime configuration isn't used by your
 * deployment of the platform.
 *
 * Exporting a config object:
 * ```
 * const config = {
 *   LMS_BASE_URL: 'http://localhost:18000'
 * };
 *
 * export default config;
 * ```
 *
 * Exporting a function that returns an object:
 * ```
 * function getConfig() {
 *   return {
 *     LMS_BASE_URL: 'http://localhost:18000'
 *   };
 * }
 * ```
 *
 * Exporting a function that returns a promise that resolves to an object:
 * ```
 * function getAsyncConfig() {
 *   return new Promise((resolve, reject) => {
 *     resolve({
 *       LMS_BASE_URL: 'http://localhost:18000'
 *     });
 *   });
 * }
 *
 * export default getAsyncConfig;
 * ```
 *
 * ##### Initialization Config Handler
 *
 * The configuration document can be extended by
 * applications at run-time using a `config` initialization handler.  Please see the Initialization
 * documentation for more information on handlers and initialization phases.
 *
 * ```
 * initialize({
 *   handlers: {
 *     config: () => {
 *       mergeConfig({
 *         CUSTOM_VARIABLE: 'custom value',
 *         LMS_BASE_URL: 'http://localhost:18001' // You can override variables, but this is uncommon.
  *       }, 'App config override handler');
 *     },
 *   },
 * });
 * ```
 *
 * ##### Runtime Configuration
 *
 * Configuration variables can also be supplied using the "runtime configuration" method, taking
 * advantage of the Micro-frontend Config API in edx-platform. More information on this API can be
 * found in the ADR which introduced it:
 *
 * https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/mfe_config_api/docs/decisions/0001-mfe-config-api.rst
 *
 * The runtime configuration method can be enabled by supplying a MFE_CONFIG_API_URL via one of the other
 * two configuration methods above.
 *
 * Runtime configuration is particularly useful if you need to supply different configurations to
 * a single deployment of a micro-frontend, for instance.  It is also a perfectly valid alternative
 * to build-time configuration, though it introduces an additional API call to edx-platform on MFE
 * initialization.
 *
 *
 * @module Config
 */

import merge from 'lodash.merge';
import {
  App,
  EnvironmentTypes,
  OptionalSiteConfig,
  Remote,
  RequiredSiteConfig,
  SiteConfig
} from '../../types';
import { APPS_CHANGED, CONFIG_CHANGED } from '../constants';
import { publish } from '../subscriptions';

let config: SiteConfig = {
  ACCESS_TOKEN_COOKIE_NAME: 'edx-jwt-cookie-header-payload',
  CSRF_TOKEN_API_PATH: '/csrf/api/v1/token',
  ENVIRONMENT: EnvironmentTypes.PRODUCTION,
  IGNORED_ERROR_REGEX: null,
  LANGUAGE_PREFERENCE_COOKIE_NAME: 'openedx-language-preference',
  PUBLIC_PATH: '/',
  REFRESH_ACCESS_TOKEN_API_PATH: '/login_refresh',
  USER_INFO_COOKIE_NAME: 'edx-user-info',
  MFE_CONFIG_API_URL: null,

  SEGMENT_KEY: null,

  apps: [],
  remotes: [],
  federatedApps: [],
  externalRoutes: [],

  pluginSlots: {},
  custom: {},

  APP_ID: '',
  BASE_URL: '',
  SITE_NAME: '',

  // Frontends
  LOGIN_URL: '',
  LOGOUT_URL: '',

  // Backends
  LMS_BASE_URL: '',

  // Branding
  FAVICON_URL: '',
  LOGO_TRADEMARK_URL: '',
  LOGO_URL: '',
  LOGO_WHITE_URL: ''
};

/**
 * Getter for the application configuration document.  This is synchronous and merely returns a
 * reference to an existing object, and is thus safe to call as often as desired.
 *
 * Example:
 *
 * ```
 * import { getConfig } from '@openedx/frontend-base';
 *
 * const {
 *   LMS_BASE_URL,
 * } = getConfig();
 * ```
 *
 * @returns {SiteConfig}
  */
export function getConfig() {
  return config;
}

/**
 * Replaces the existing SiteConfig.  This is not commonly used, but can be helpful for tests.
 *
 * Example:
 *
 * ```
 * import { setConfig } from '@openedx/frontend-base';
 *
 * setConfig({
 *   LMS_BASE_URL, // This is overriding the ENTIRE document - this is not merged in!
 * });
 * ```
 *
 * @param newConfig A replacement SiteConfig which will completely override the current SiteConfig.
 */
export function setConfig(newConfig: SiteConfig) {
  config = newConfig;
  publish(CONFIG_CHANGED);
}

/**
 * Merges additional configuration values into the site config returned by `getConfig`.  Will
 * override any values that exist with the same keys.
 *
 * ```
 * mergeConfig({
 *   NEW_KEY: 'new value',
 *   OTHER_NEW_KEY: 'other new value',
 * });
 *
 * This function uses lodash.merge internally to merge configuration objects
 * which means they will be merged recursively.  See https://lodash.com/docs/latest#merge for
 * documentation on the exact behavior.
 *
 * @param {Object} newConfig
 */
export function mergeConfig(newConfig: Partial<Partial<OptionalSiteConfig> & RequiredSiteConfig>) {
  config = merge(config, newConfig);
  publish(CONFIG_CHANGED);
}

export function patchApp(app: App) {
  config.apps.push(app);
  publish(APPS_CHANGED);
  publish(CONFIG_CHANGED);
}

/**
 * This function will attempt to add the supplied remotes to the SiteConfig.  If a remote with that
 * ID already exists, it will not be added.  If changes are made to the SiteConfig, the
 * CONFIG_CHANGED event will be published.
 *
 * @param remotes An array of Remote objects to merge into the SiteConfig's `remotes` array.
 */
export function mergeRemotes(remotes: Remote[]) {
  const remoteIds = config.remotes.map(remote => remote.id);
  let changed = false;
  for (const remote of remotes) {
    if (!remoteIds.includes(remote.id)) {
      config.remotes.push(remote);
      changed = true;
    }
  }
  if (changed) {
    publish(CONFIG_CHANGED);
  }
}
