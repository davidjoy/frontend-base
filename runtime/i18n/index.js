/**
 * #### Import members from **@openedx/frontend-base**
 * The i18n module relies on react-intl and re-exports all of that package's exports.
 *
 * For each locale we want to support, react-intl needs 1) the locale-data, which includes
 * information about how to format numbers, handle plurals, etc., and 2) the translations, as an
 * object holding message id / translated string pairs.  A locale string and the messages object are
 * passed into the IntlProvider element that wraps your element hierarchy.
 *
 * Note that react-intl has no way of checking if the translations you give it actually have
 * anything to do with the locale you pass it; it will happily use whatever messages object you pass
 * in.  However, if the locale data for the locale you passed into the IntlProvider was not
 * correctly installed with addLocaleData, all of your translations will fall back to the default
 * (in our case English), *even if you gave IntlProvider the correct messages object for that
 * locale*.
 *
 * Messages are provided to this module via the configureI18n() function below.
 *
 *
 * @module Internationalization
 * @see {@link https://github.com/openedx/frontend-base/blob/master/docs/how_tos/i18n.rst}
 * @see {@link https://formatjs.io/docs/react-intl/components/ Intl} for components exported from this module.
 *
 */

/**
 * @name createIntl
 * @kind function
 * @see {@link https://formatjs.io/docs/react-intl/api#createIntl Intl}
 */

/**
 * @name FormattedDate
 * @kind class
 * @see {@link https://formatjs.io/docs/react-intl/components/#formatteddate Intl}
 */

/**
 * @name FormattedTime
 * @kind class
 * @see {@link https://formatjs.io/docs/react-intl/components/#formattedtime Intl}
 */

/**
 * @name FormattedRelativeTime
 * @kind class
 * @see {@link https://formatjs.io/docs/react-intl/components/#formattedrelativetime Intl}
 */

/**
 * @name FormattedNumber
 * @kind class
 * @see {@link https://formatjs.io/docs/react-intl/components/#formattednumber Intl}
 */

/**
 * @name FormattedPlural
 * @kind class
 * @see {@link https://formatjs.io/docs/react-intl/components/#formattedplural Intl}
 */

/**
 * @name FormattedMessage
 * @kind class
 * @see {@link https://formatjs.io/docs/react-intl/components/#formattedmessage Intl}
 */

/**
 * @name IntlProvider
 * @kind class
 * @see {@link https://formatjs.io/docs/react-intl/components/#intlprovider Intl}
 */

/**
 * @name defineMessages
 * @kind function
 * @see {@link https://formatjs.io/docs/react-intl/api#definemessagesdefinemessage Intl}
 */

/**
 * @name useIntl
 * @kind function
 * @see {@link https://formatjs.io/docs/react-intl/api#useIntl Intl}
 */

export {
  createIntl,
  defineMessages,
  FormattedDate,
  FormattedMessage,
  FormattedNumber,
  FormattedPlural,
  FormattedRelativeTime,
  FormattedTime,
  IntlProvider,
  useIntl
} from 'react-intl';

export {
  addAppMessages,
  configureI18n,
  getLocale,
  getLocalizedLanguageName,
  getMessages,
  getPrimaryLanguageSubtag,
  getSupportedLanguageList,
  handleRtl,
  intlShape,
  isRtl,
  LOCALE_CHANGED,
  LOCALE_TOPIC,
  mergeMessages,
  updateLocale
} from './lib';

export {
  default as injectIntl
} from './injectIntlWithShim';
