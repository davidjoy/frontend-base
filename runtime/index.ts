export {
  MockAnalyticsService,
  SegmentAnalyticsService,
  configure as configureAnalytics,
  getAnalyticsService,
  identifyAnonymousUser,
  identifyAuthenticatedUser,
  resetAnalyticsService,
  sendPageEvent,
  sendTrackEvent,
  sendTrackingLogEvent
} from './analytics';

export {
  AUTHENTICATED_USER_CHANGED,
  AUTHENTICATED_USER_TOPIC,
  AxiosJwtAuthService,
  MockAuthService,
  configure as configureAuth,
  ensureAuthenticatedUser,
  fetchAuthenticatedUser,
  getAuthService,
  getAuthenticatedHttpClient,
  getAuthenticatedUser,
  getHttpClient,
  getLoginRedirectUrl,
  getLogoutRedirectUrl,
  hydrateAuthenticatedUser,
  redirectToLogin,
  redirectToLogout,
  setAuthenticatedUser
} from './auth';

export {
  getConfig,
  mergeConfig,
  setConfig
} from './config';

export {
  APP_ANALYTICS_INITIALIZED,
  APP_AUTH_INITIALIZED,
  APP_CONFIG_INITIALIZED,
  APP_I18N_INITIALIZED,
  APP_INIT_ERROR,
  APP_LOGGING_INITIALIZED,
  APP_PUBSUB_INITIALIZED,
  APP_READY,
  APP_TOPIC,
  CONFIG_CHANGED,
  CONFIG_TOPIC
} from './constants';

export {
  FormattedDate,
  FormattedMessage,
  FormattedNumber,
  FormattedPlural,
  FormattedRelativeTime,
  FormattedTime,
  IntlProvider,
  LOCALE_CHANGED,
  LOCALE_TOPIC,
  configure as configureI18n,
  createIntl,
  defineMessages,
  getLocale,
  getLocalizedLanguageName,
  getMessages,
  getPrimaryLanguageSubtag,
  getSupportedLanguageList,
  handleRtl,
  injectIntl,
  intlShape,
  isRtl,
  patchMessages,
  updateLocale,
  useIntl
} from './i18n';

export {
  auth,
  getBasename,
  getHistory,
  initError,
  initialize
} from './initialize';

export {
  MockLoggingService,
  NewRelicLoggingService,
  configure as configureLogging,
  getLoggingService,
  logError,
  logInfo,
  resetLoggingService
} from './logging';

export {
  Plugin,
  PluginSlot
} from './plugins';

export {
  AppContext,
  AppProvider,
  AuthenticatedPageRoute,
  Divider,
  ErrorBoundary,
  ErrorPage,
  LoginRedirect,
  PageWrap,
  useAppEvent,
  useAuthenticatedUser,
  useConfig
} from './react';

export {
  clearAllSubscriptions,
  publish,
  subscribe,
  unsubscribe
} from './subscriptions';

export {
  initializeMockApp,
  mockMessages
} from './testing';

export {
  camelCaseObject,
  convertKeyNames,
  getPath,
  getQueryParameters,
  modifyObjectKeys,
  parseURL,
  snakeCaseObject
} from './utils';
