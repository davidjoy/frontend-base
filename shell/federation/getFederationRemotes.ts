import { getConfig, isValidVariableName } from '../../runtime';

export function getFederationRemotes() {
  const { remotes } = getConfig();
  if (Array.isArray(remotes)) {
    // Validate the remote IDs are valid names.
    remotes.forEach(remote => {
      if (!isValidVariableName(remote.id)) {
        throw new Error(`Module federation error.\n\nThe remote ID "${remote.id}" is invalid. This remote's URL is "${remote.url}".\n\nThe identifier must be a valid JavaScript variable name.  It must start with a letter, cannot be a reserved word, and can only contain letters, digits, underscores and dollar signs.`);
      }
    });

    return remotes.map((remote) => ({
      name: remote.id,
      // We add a date here to ensure that we cache bust the remote entry file regardless of what
      // headers it returns to us.  This ensures that even if operators haven't set up their
      // caching headers correctly, we always get the most recent version.
      entry: `${remote.url}?${new Date().getTime()}`,
    }));
  }
  return [];
}
