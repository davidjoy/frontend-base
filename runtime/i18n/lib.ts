import merge from 'lodash.merge';
import PropTypes from 'prop-types';
import { MessageFormatElement } from 'react-intl';
import Cookies from 'universal-cookie';

import { getConfig } from '../config';
import { publish } from '../subscriptions';

const cookies = new Cookies();
const rtlLocales = [
  'ar', // Arabic
  'he', // Hebrew
  'fa', // Farsi (not currently supported)
  'fa-ir', // Farsi Iran
  'ur', // Urdu (not currently supported)
];

let messages: Record<string, Record<string, string> | Record<string, MessageFormatElement[]> | undefined>;

/**
 * @memberof module:Internationalization
 *
 * Prior versions of react-intl (our primary implementation of the i18n service) included a
 * PropTypes-based 'shape' for its `intl` object.  This has since been removed.  For legacy
 * compatibility, we include an `intlShape` export that is set to PropTypes.object.  Usage of this
 * export is deprecated.
 *
 * @deprecated
 */
export const intlShape = PropTypes.object;

/**
 * @memberof module:Internationalization
 */
export const LOCALE_TOPIC = 'LOCALE';

/**
 * @memberof module:Internationalization
 */
export const LOCALE_CHANGED = `${LOCALE_TOPIC}.CHANGED`;

/**
 *
 * @memberof module:Internationalization
 * @returns {Cookies}
 */
export function getCookies() {
  return cookies;
}

/**
 * Some of our dependencies function on primary language subtags, rather than full locales.
 * This function strips a locale down to that first subtag.  Depending on the code, this
 * may be 2 or more characters.
 *
 * @param {string} code
 * @memberof module:Internationalization
 */
export function getPrimaryLanguageSubtag(code) {
  return code.split('-')[0];
}

/**
 * Finds the closest supported locale to the one provided.  This is done in three steps:
 *
 * 1. Returning the locale itself if its exact language code is supported.
 * 2. Returning the primary language subtag of the language code if it is supported (ar for ar-eg,
 * for instance).
 * 3. Returning 'en' if neither of the above produce a supported locale.
 *
 * @param {string} locale
 * @returns {string}
 * @memberof module:Internationalization
 */
export function findSupportedLocale(locale) {
  if (messages === undefined) {
    throw new Error('findSupportedLocale called before configuring i18n. Call configure with messages first.');
  }

  if (messages[locale] !== undefined) {
    return locale;
  }

  if (messages[getPrimaryLanguageSubtag(locale)] !== undefined) {
    return getPrimaryLanguageSubtag(locale);
  }

  return 'en';
}

/**
 * Get the locale from the cookie or, failing that, the browser setting.
 * Gracefully fall back to a more general primary language subtag or to English (en)
 * if we don't support that language.
 *
 * @param {string|undefined} locale If a locale is provided, returns the closest supported locale. Optional.
 * @throws An error if i18n has not yet been configured.
 * @returns {string}
 * @memberof module:Internationalization
 */
export function getLocale(locale?: string) {
  if (messages === null) {
    throw new Error('getLocale called before configuring i18n. Call configure with messages first.');
  }

  // 1. Explicit application request
  if (locale !== undefined) {
    return findSupportedLocale(locale);
  }
  // 2. User setting in cookie

  const cookieLangPref = cookies.get(getConfig().LANGUAGE_PREFERENCE_COOKIE_NAME);
  if (cookieLangPref) {
    return findSupportedLocale(cookieLangPref.toLowerCase());
  }
  // 3. Browser language (default)
  // Note that some browers prefer upper case for the region part of the locale, while others don't.
  // Thus the toLowerCase, for consistency.
  // https://developer.mozilla.org/en-US/docs/Web/API/NavigatorLanguage/language
  return findSupportedLocale(globalThis.navigator.language.toLowerCase());
}

export function getLocalizedLanguageName(locale) {
  const localizedName = (new Intl.DisplayNames([locale], { type: 'language' })).of(locale);

  if (localizedName === undefined) {
    throw new Error(`Unsupported locale: ${locale}`);
  }

  return `${localizedName.charAt(0).toLocaleUpperCase(locale)}${localizedName.slice(1)}`;
}

export function getSupportedLanguageList() {
  const locales = Object.keys(messages);
  locales.push('en'); // 'en' is not in the messages object because it's the default.
  locales.sort();

  return locales.map((locale) => ({
    code: locale,
    name: getLocalizedLanguageName(locale),
  }));
}

export function updateLocale() {
  handleRtl();
  publish(LOCALE_CHANGED);
}

/**
 * Returns messages for the provided locale, or the user's preferred locale if no argument is
 * provided.
 *
 * @param {string} [locale=getLocale()]
 * @memberof module:Internationalization
 */
export function getMessages(locale = getLocale()) {
  if (messages === undefined) {
    throw new Error('getMessages called before configuring i18n. Call configure with messages first.');
  }

  return messages[locale];
}

/**
 * Determines if the provided locale is a right-to-left language.
 *
 * @param {string} locale
 * @memberof module:Internationalization
 */
export function isRtl(locale) {
  return rtlLocales.includes(locale);
}

/**
 * Handles applying the RTL stylesheet and "dir=rtl" attribute to the html tag if the current locale
 * is a RTL language.
 *
 * @memberof module:Internationalization
 */
export function handleRtl() {
  if (isRtl(getLocale())) {
    globalThis.document.getElementsByTagName('html')[0].setAttribute('dir', 'rtl');
  } else {
    globalThis.document.getElementsByTagName('html')[0].setAttribute('dir', 'ltr');
  }
}

/**
 *
 *
 * @param {Object} newMessages
 * @returns {Object}
 * @memberof module:Internationalization
 */
export function mergeMessages(newMessages) {
  const msgs = Array.isArray(newMessages) ? merge({}, ...newMessages) : newMessages;
  messages = merge(messages, msgs);

  return messages;
}

interface ConfigureI18nOptions {
  messages: Record<string, Record<string, string>>[] | Record<string, Record<string, string>>,
}

/**
 * Configures the i18n library with messages for your application.
 *
 * Logs a warning if it detects a locale it doesn't expect (as defined by the supportedLocales list
 * above), or if an expected locale is not provided.
 *
 * @param {Object} options
 * @param {Object} options.messages
 * @memberof module:Internationalization
 */
export function configure(options: ConfigureI18nOptions) {
  messages = Array.isArray(options.messages) ? merge({}, ...options.messages) : options.messages;

  handleRtl();
}
