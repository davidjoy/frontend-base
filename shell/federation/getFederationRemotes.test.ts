import { mergeConfig } from '../../runtime';
import { getFederationRemotes } from './getFederationRemotes';

describe('getFederationRemotes', () => {
  let getTimeOriginal: () => number = () => {
    return 123;
  };
  beforeEach(() => {
    getTimeOriginal = Date.prototype.getTime;
    Date.prototype.getTime = jest.fn(() => 123);
  });

  afterEach(() => {
    Date.prototype.getTime = getTimeOriginal;
  });

  it('should return an empty array when there are no remotes', () => {
    const remotes = getFederationRemotes();
    expect(remotes).toEqual([]);
  });

  it('should return a formatted remotes when there are remotes', () => {
    const remotes = [
      {
        id: 'remote1',
        url: 'http://localhost:8001/remoteEntry.js',
      },
      {
        id: 'remote2',
        url: 'http://localhost:8002/remoteEntry.js',
      }
    ];
    mergeConfig({
      remotes
    });

    const result = getFederationRemotes();
    expect(result).toHaveLength(2);
    expect(result[0]).toEqual({
      name: 'remote1',
      entry: 'http://localhost:8001/remoteEntry.js?123'
    });
    expect(result[1]).toEqual({
      name: 'remote2',
      entry: 'http://localhost:8002/remoteEntry.js?123'
    });
  });

  it('should throw an error if a remote ID is not a valid variable name', () => {
    const remotes = [
      {
        id: 'remote1',
        url: 'http://localhost:8001/remoteEntry.js',
      },
      {
        id: 'remote-2', // This is an invalid name because it contains a '-'
        url: 'http://localhost:8002/remoteEntry.js',
      }
    ];
    mergeConfig({
      remotes
    });

    expect(
      () => getFederationRemotes()
    ).toThrowErrorMatchingSnapshot();
  });
});
