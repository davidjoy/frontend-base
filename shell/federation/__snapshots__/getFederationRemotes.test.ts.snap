// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFederationRemotes should throw an error if a remote ID is not a valid variable name 1`] = `
"Module federation error.

The remote ID "remote-2" is invalid. This remote's URL is "http://localhost:8002/remoteEntry.js".

The identifier must be a valid JavaScript variable name.  It must start with a letter, cannot be a reserved word, and can only contain letters, digits, underscores and dollar signs."
`;
